src/machi_file_proxy.erl

```diff
@@ -729,3 +796,43 @@ maybe_regenerate_checksum(FHd, {Offset, Size, _Csum}) ->
         Error ->
             throw(Error)
     end.
+
+%% GC: make sure unwritte bytes = [{Eof, infinity}] and Eof is > max file size
+%% walk through the checksum table and make sure all chunks trimmed
+%% Then unlink the file
+-spec maybe_gc(term(), #state{}) ->
+                      {reply, term(), #state{}} | {stop, normal, term(), #state{}}.
+maybe_gc(Reply, S = #state{eof_position = Eof,
+                           max_file_size = MaxFileSize}) when Eof < MaxFileSize ->
+    lager:debug("The file is still small; not trying GC (Eof, MaxFileSize) = (~p, ~p)~n",
+                [Eof, MaxFileSize]),
+    {reply, Reply, S};
+maybe_gc(Reply, S = #state{fluname=FluName,
+                           data_filehandle = FHd,
+                           data_dir = DataDir,
+                           filename = Filename,
+                           eof_position = Eof,
+                           csum_table=CsumTable}) ->
+    case machi_csum_table:all_trimmed(CsumTable, Eof) of
+        true ->
+            lager:debug("GC? Let's do it: ~p.~n", [Filename]),
+            %% Before unlinking a file, it should inform
+            %% machi_flu_filename_mgr that this file is
+            %% deleted and mark it as "trimmed" to avoid
+            %% filename reuse and resurrection. Maybe garbage
+            %% will remain if a process crashed but it also
+            %% should be recovered at filename_mgr startup.
+
+            %% Also, this should be informed *before* file proxy
+            %% deletes files.
+            ok = machi_flu_metadata_mgr:trim_file(FluName, {file, Filename}),
+            ok = file:close(FHd),
+            {_, DPath} = machi_util:make_data_filename(DataDir, Filename),
+            ok = file:delete(DPath),
+            machi_csum_table:delete(CsumTable),
```

Comment:
The order of the ops above can cause problems, if the kernel panics without flushing dirty memory pages to disk. (And these writes aren't all 100% synchronous, which is another problem.)  We have work scheduled for durability features (e.g. durability/sync write requests, durable writes batching, etc.), so I don't expect perfectly durable code here.

However, even if all ops by lines 828-832 where synchronous, don't we want 832 machi_csum_table:delete() happening before 831 file:delete()?